javabuilder-authorizer/jwt_helper.rb

```diff
@@ -0,0 +1,100 @@
+module JwtHelper
+  # Verify the token with the appropriate public key (dependant on the
+  # environment the request came from), and checks the token has not
+  # expired and its issue time is not in the future.
+  def decode_token(token, origin)
+    return false unless token
+    begin
+      return JWT.decode(
+        token,
+        # Temporarily choose the key based on the client origin rather than the
+        # resource until we have environment-specific Javabuilders set up.
+        get_public_key(origin),
+        true,
+        verify_iat: true, # verify issued at time is valid
+        algorithm: 'RS256'
+      )
+    rescue JWT::ExpiredSignature, JWT::InvalidIatError
+      return false
+    end
+  end
+
+  def generate_allow(resource, decoded_token)
+    token_payload = decoded_token[0]
+    user_id = token_payload['uid']
+    issuer = token_payload['iss']
+    principal_id = "#{issuer}/#{user_id}"
+    generate_policy(principal_id, 'Allow', resource, token_payload)
+  end
+
+  def generate_deny(resource)
+    generate_policy(nil, 'Deny', resource, nil)
+  end
+
+  # Helper function to generate an IAM policy.
+  def generate_policy(principal_id, effect, resource, token_payload)
+    # Required output:
+    auth_response = {}
+    auth_response['principalId'] = principal_id
+    if effect && resource
+      policy_document = {}
+      policy_document['Version'] = '2012-10-17' # default version
+      policy_document['Statement'] = []
+      statement_one = {}
+      statement_one['Action'] = 'execute-api:Invoke' # default action
+      statement_one['Effect'] = effect
+      statement_one['Resource'] = resource
+      policy_document['Statement'][0] = statement_one
+      auth_response['policyDocument'] = policy_document
+    end
+    auth_response['context'] = token_payload if token_payload
+    auth_response
+  end
+
+  private
+
+  # Load the JWT public key from the appropriate environment variable
+  # based on the environment the request was sent from.
+  #
+  # route_arn is in format
+  # arn:aws:execute-api:region:account-id:api-id/stage-name/$connect
+  # We only care about stage--that tells us the environment (development, staging, etc.)
+  # def get_public_key(route_arn)
+  def get_public_key(origin)
+    # Temporarily choose the key based on the client origin rather than the
+    # route_arn until we have environment-specific Javabuilders set up.
+    # tmp = route_arn.split(':')
+    # api_gateway_arn = tmp[5].split('/')
+    # stage_name = api_gateway_arn[1]
+
+    # TODO: Update the following strings to use the exact origin that requests come from.
```

Comment:
do we have a work item for this task?

Reply:
Yeah, it's the second half of this task: https://codedotorg.atlassian.net/browse/JAVA-471